8290059: Do not use std::thread in panama tests

JornVernee opened the pull request:
This patch removes the use of std::thread from the `java.lang.foreign` tests, and switches to the OS specific thread APIs, in order to change things such as the stack size on some platforms where this is required in the future (see the JBS issue).

This is done by adding a small header-only library that exposes a function to run code in freshly spawned threads (`run_async`).

Testing: Running the affected tests on platforms that implement the linker.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290059](https://bugs.openjdk.org/browse/JDK-8290059): Do not use std::thread in panama tests


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [49e60f2d](https://git.openjdk.org/jdk/pull/9599/files/49e60f2d187476683d32fcab02714a4c56315529)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [49e60f2d](https://git.openjdk.org/jdk/pull/9599/files/49e60f2d187476683d32fcab02714a4c56315529)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [49e60f2d](https://git.openjdk.org/jdk/pull/9599/files/49e60f2d187476683d32fcab02714a4c56315529)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9599/head:pull/9599` \
`$ git checkout pull/9599`

Update a local copy of the PR: \
`$ git checkout pull/9599` \
`$ git pull https://git.openjdk.org/jdk pull/9599/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9599`

View PR using the GUI difftool: \
`$ git pr show -t 9599`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9599.diff">https://git.openjdk.org/jdk/pull/9599.diff</a>

</details>

mcimadamore reviewed (approved):
Looks good!

mcimadamore commented on test/lib/native/testlib_threads.h:
```diff
@@ -0,0 +1,89 @@
+#include <windows.h>
+typedef HANDLE THREAD_ID;
+
+#else // !_WIN32
```
the comments in the #else are inconsistent - some use `!_WIN32`, some use `!windows` others use nothing.

RealCLanger commented:
@JornVernee thanks for doing this so quickly. I suggest undoing https://github.com/openjdk/jdk19/commit/c29242ebb0cfd3eaa56240664af607c02a11324e in this change, too. If you update this PR I can run it through our Alpine test pipeline.

JornVernee commented:
@RealCLanger Note that I'm not setting the stack size of the thread in this patch. I'm just using the defaults. From the discussion on the bug, I don't think this sufficient to make the tests pass on Apline/MuslC.

I avoided getting into that since I don't have ready access to an Alpine/MuslC testing environment atm. So, I've left setting the stack size on MuslC, and re-enabling the tests for someone that does. Hopefully this patch is enough to get that going easily.

JornVernee commented:
@mcimadamore Thanks for the comments. I've cleaned of the comments.

I also realized there was an unused typedef for `THREAD_ID`, which I've removed. Finally, I realized that after printing an error, we should probably also exit the test (I'd forgotten to add this before making the PR public). I've added a helper function called `fatal` that prints the error message and then calls `exit`. The tests still pass with that.

RealCLanger commented:
> @RealCLanger Note that I'm not setting the stack size of the thread in this patch. I'm just using the defaults. From the discussion on the bug, I don't think this sufficient to make the tests pass on Apline/MuslC.
> 
> I avoided getting into that since I don't have ready access to an Alpine/MuslC testing environment atm. So, I've left setting the stack size on MuslC, and re-enabling the tests for someone that does. Hopefully this patch is enough to get that going easily.

OK, thanks for clarifying that. @tstuefe, maybe you want to have a look after this fix is in?

tstuefe commented on test/lib/native/testlib_threads.h:
```diff
@@ -0,0 +1,89 @@
+
+static void fatal(const char* message) {
+    perror(message);
+    exit(-1);
```
Won't work as intended for Windows APIs. I would print the result of `GetLastError()` instead. I would be fine with just omitting the error code too, because I think the old tests did not handle errors either. Or did we catch std::thread exceptions somewhere?

JornVernee replied:
The intent was to exit the test with a non-zero exit code, in order to avoid any accidental false positives.

I could return the error code from `GetLastError` and from the respective pthread apis as an exit code instead. Is that what you mean?

JornVernee replied:
FWIW, `perror` just prints to `stderr`: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/perror-wperror?view=msvc-170

JornVernee replied:
Ah, I see what you mean now. The C standard library function `perror` defines this to also print a textual description of `errno`. https://en.cppreference.com/w/c/io/perror so that won't work for the Windows APIs.

I think the alternative would be to use `FormatMessage` on Windows. I didn't really think that much into this.

Maybe it's clearer to use `fputs` with `stderr` here.

tstuefe commented on test/lib/native/testlib_threads.h:
```diff
@@ -0,0 +1,89 @@
+    Helper* helper = (Helper*)ctxt;
+    helper->proc(helper->context);
+    return 0;
+}
```
I'm curious, does this only exist because of the DWORD vs void* return value size of the native thread functions? I wondered why you don't just pass the test procedure to the thread start API directly.

About stdcall, does Oracle still build Windows 32bit?

JornVernee replied:
Yes, this exists as an adapter from PROCEDURE to whatever callback type the OS API expects. I tried passing the procedure to the start API as well, but the compiler complains that the types don't match. Even if I make PROCEDURE return `int`, it seems to want `DWORD` explicitly.

This was taken from the old library code which uses `_beginthreadex`. That one explicitly wants  to use `__stdcall`, so I just matched that. Looking again, it doesn't look like the same is required for `CreateThread` [1]. I'll remove it for clarity.

[1] : https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms686736(v=vs.85)

tstuefe reviewed (approved):
Hi @JornVernee,

good job and thanks for doing it so quickly. Looks good, module the GetLastError comment.

About the stack size and Alpine, I am fine with doing the Aĺpine-specific fix in a follow-up, or you can do it as part of this change. Strictly speaking its not part of this bug report to adjust the stack size.

If you fix the GetLastError issue, I don't need another look.
Thanks, Thomas

dholmes-ora commented:
`run_async` is an odd name for this, especially as the fact you create then join mean it doesn't run asynchronously at all - it runs synchronously in another thread. The API would be more generally useful if you split the start and the join, but that would require also exposing an opaque thread "handle" (though splitting also means you don't have to try and think of a good name for `run_in_a_new_thread-and_join_it()` - you just have `threadStart` and `threadJoin`).

erikj79 reviewed (approved):
Build changes look good.

JornVernee commented:
> `run_async` is an odd name for this, especially as the fact you create then join mean it doesn't run asynchronously at all - it runs synchronously in another thread. 

I took the name from the CompleteableFuture API [1], although in that case a future is returned on which a user can call `get`. I can rename the function. Does `run_in_new_thread` seem good enough?

> The API would be more generally useful if you split the start and the join, but that would require also exposing an opaque thread "handle" (though splitting also means you don't have to try and think of a good name for `run_in_a_new_thread-and_join_it()` - you just have `threadStart` and `threadJoin`).

I can see this being useful in order to start many threads up front, and then join then all afterwards (to get them running concurrently). Though, I'd like to hold off on getting into that since the current tests don't need that functionality. The library header can be freely amended later as well.

[1] : https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/util/concurrent/CompletableFuture.html#runAsync(java.lang.Runnable)

JornVernee commented:
I made some tentative changes based on the review comments:

- Added some code comments.
- Switch `perror` to `fputs` with `stderr` for clarity.
- Use the OS API error code as exit code.
- Switch the windows type of `procedure` to exactly what the API expects: https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms686736(v=vs.85)
- Renamed `run_async` to `run_in_new_thread`.

Please take another look.

dholmes-ora commented:
> Does run_in_new_thread seem good enough?

No, sorry, the fact it  both runs and joins is a critical aspect. `run_async` in `CompleteableFuture` just does the "run in new thread" part, whereas the `get()` on the returned `FutureTask` provides the "join". So a function that does both needs to make that clear in the name - and there is no nice succinct name for that, so we get stuck with something like `run_in_new_thread_and_join`. Or we just have ` startThread` and `joinThread` :)

JornVernee commented:
> > Does run_in_new_thread seem good enough?
> 
> No, sorry, the fact it both runs and joins is a critical aspect. `run_async` in `CompleteableFuture` just does the "run in new thread" part, whereas the `get()` on the returned `FutureTask` provides the "join". So a function that does both needs to make that clear in the name - and there is no nice succinct name for that, so we get stuck with something like `run_in_new_thread_and_join`. Or we just have ` startThread` and `joinThread` :)

`startThread` and `joinThread` is not the functionality that the tests need. The functionality that's needed is one that combines these two, so that's what I went with. An API without real users is harder to get right.

I've looked into adding `startThread` and `joinThread`. But, the story is not so simple. Since now the context passed to `CreateThread` and `pthread_create` has to outlive the function, which means heap allocating, or asking the caller to stack allocate. I think a typical C API might have a `new_thread` and `delete_thread` to do the memory management. These functions could be folded into `startThread` and `joinThread`, and we'd rename those to `create_and_start_thread` and `join_and_destroy_thread` (I think it's important to signal to callers that allocation happens, so that the memory will be cleaned up by calling join as well), but at that point it feels like we're back where we started. The alternative being stack allocation. i.e. the caller declares a `THREAD` variable and passes a pointer to it to `start_thread`, which then uses that memory. This avoids the heap allocation, but adds more noise in the caller.

Getting into that just feels like too much unneeded complexity at this moment. So, `run_in_new_thread_and_join` it is.

dholmes-ora commented:
> startThread and joinThread is not the functionality that the tests need.

But it is the functionality the tests are already using.

> now the context passed to CreateThread and pthread_create has to outlive the function

That's not an issue for the API to solve, that is an issue for the user of the API to solve. pthread_create doesn't involve any memory management.

> So, run_in_new_thread_and_join it is.

Okay, that is fine.

JornVernee commented:
@tstuefe Do the changes I made look good? If so I'll go ahead and integrate.

tstuefe commented:
Looks still good.

JornVernee commented:
/integrate
